### PR TITLE
feat: added `debug` flag in the SuperTokenConfig logging

### PR DESCRIPTION
**Pull request by IamMayankThakur**
## Summary of change

Added debug flag in the SuperTokenConfig in the init() for logging

## Related issues

-   Link to issue1 here
-   Link to issue1 here

## Test Plan

Added unit tests

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [x] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens/constants.go`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [x] Changes to the version if needed
    -   In `supertokens/constants.go > version variable`
-   [x] Had installed and ran the pre-commit hook
-   [ ] If new thirdparty provider is added,
    -   [ ] update switch statement in `recipe/thirdparty/providers/config_utils.go` file, `createProvider` function.
    -   [ ] add an icon on the user management dashboard.
-   [x] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If access token structure has changed
    -   Modified test in `session/accessTokenVersions_test.go` to account for any new claims that are optional or omitted by the core

### CODE REVIEW

File: supertokens/models.go [hunk 1]
```diff
@@ -46,6 +46,7 @@ type TypeInput struct {
 	AppInfo               AppInfo
 	RecipeList            []Recipe
 	Telemetry             *bool
+	Debug                 *bool
```
Comment (rishabhpoddar) [inline supertokens/models.go:49]:
```suggestion
	Debug                 bool
```

File: supertokens/supertokens.go [hunk 2]
```diff
@@ -35,6 +36,7 @@ type superTokens struct {
 	RecipeModules         []RecipeModule
 	OnSuperTokensAPIError func(err error, req *http.Request, res http.ResponseWriter)
 	Telemetry             *bool
+	Debug                 *bool
```
Comment (rishabhpoddar) [inline supertokens/supertokens.go:39]:
```suggestion
	Debug                 bool
```

File: supertokens/supertokens.go [hunk 3]
```diff
@@ -54,6 +56,14 @@ func supertokensInit(config TypeInput) error {
 		superTokens.OnSuperTokensAPIError = config.OnSuperTokensAPIError
 	}
 
+	superTokens.Debug = config.Debug
+	if superTokens.Debug != nil && *superTokens.Debug == true {
```
Comment (rishabhpoddar) [inline supertokens/supertokens.go:60]:
```suggestion
	if superTokens.Debug == true {
```

File: supertokens/supertokens.go [hunk 4]
```diff
@@ -54,6 +56,14 @@ func supertokensInit(config TypeInput) error {
...
+		err := os.Setenv("SUPERTOKENS_DEBUG", "1")
```
Comment (rishabhpoddar) [inline supertokens/supertokens.go:61]: this should not be done this way. The logger should read the boolean

File: recipe/session/logger_test.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	"bytes"
+	"log"
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/supertokens/supertokens-golang/supertokens"
+	"github.com/supertokens/supertokens-golang/test/unittesting"
+)
+
```
Comment (rishabhpoddar) [inline recipe/session/logger_test.go:13]: add a comment explaining why this file is here

File: recipe/session/logger_test.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+func TestLogDebugMessageWhenDebugTrue(t *testing.T) {
+	var logMessage = "test log message"
+	var buf bytes.Buffer
+
+	debug := true
+	supertokens.Logger = log.New(&buf, "", 0)
```
Comment (rishabhpoddar) [inline recipe/session/logger_test.go:19]: you need to reset this to the original value after each test

File: recipe/session/logger_test.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	assert.Equal(t, &debug, supertokensInstance.Debug)
+	assert.NotContains(t, buf.String(), logMessage, "checking log message in logs")
+}
+
+func TestLogDebugMessageWithEnvVar(t *testing.T) {
+	var logMessage = "test log message"
+	var buf bytes.Buffer
+
+	supertokens.Logger = log.New(&buf, "", 0)
+	os.Setenv("SUPERTOKENS_DEBUG", "1")
```
Comment (rishabhpoddar) [inline recipe/session/logger_test.go:107]: remove the setting of the env after this test

File: recipe/session/logger_test.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+	if err != nil {
+		t.Error(err.Error())
+	}
+
+	supertokens.LogDebugMessage(logMessage)
+	assert.Equal(t, &debug, supertokensInstance.Debug)
+	assert.NotContains(t, buf.String(), logMessage, "checking log message in logs")
+}
+
+func TestLogDebugMessageWithEnvVar(t *testing.T) {
```
Comment (rishabhpoddar) [inline recipe/session/logger_test.go:102]: add another test where you don't have the env and don't have the Debug set, and make sure no logs are happening

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: recipe/session/testingUtils.go [hunk 9]
```diff
@@ -48,6 +50,11 @@ func AfterEach() {
 	unittesting.CleanST()
 }
 
+func resetLogger() {
```
Comment (rishabhpoddar) [inline recipe/session/testingUtils.go:53]: move this function to the logger_test file.. since it's only ever used in that

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens/supertokens.go [hunk 10]
```diff
@@ -54,6 +55,9 @@ func supertokensInit(config TypeInput) error {
 		superTokens.OnSuperTokensAPIError = config.OnSuperTokensAPIError
 	}
 
+	superTokens.Debug = config.Debug
+	superTokensInstance = superTokens
```
Comment (rishabhpoddar) [inline supertokens/supertokens.go:59]: why is this here?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens/supertokens.go (hunk 10 above)
Comment (IamMayankThakur) [reply supertokens/supertokens.go:59]: Because `superTokensInstance` is initialised only at the end of `supertokensInit()`, but its needed in the `logDebugMessage()` function.

File: recipe/session/logger_test.go [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/supertokens/supertokens-golang/supertokens"
+	"github.com/supertokens/supertokens-golang/test/unittesting"
+)
+
+// Added the logger tests here because supertokens/logger_test.go causes cyclic import errors due to imports in test/unittesting/testingUtils.go
+
+func resetLogger() {
+	supertokens.Logger = log.New(os.Stdout, "com.supertokens", 0)
+	os.Unsetenv("SUPERTOKENS_DEBUG")
```
Comment (rishabhpoddar) [inline recipe/session/logger_test.go:18]: shouldnt debugEnabled be reset to false after the test?

Comment (IamMayankThakur) [reply recipe/session/logger_test.go:18]: Done
